modules/internal/DefaultAssociative.chpl

```diff
@@ -566,14 +566,15 @@ module DefaultAssociative {
 
     // ref version
     proc dsiAccess(idx : idxType, haveLock = false) ref {
-      const shouldLock = dom.parSafe && !haveLock;
+      const arrOwnsDom = dom._arrs.length == 1;
+      const shouldLock = dom.parSafe && !haveLock && arrOwnsDom;
```

Comment:
As I understand it, the locking is used to ensure that the domain resize operation happens under a lock. What if this dsiAccess first tried to get the element without locking, and then if it is not found try again with locking? (Noting that on the second try, the element might have appeared in the meantime)? That way, the only time the lock is invoked on element access is when actually adding an index to the domain.